File: src/main/scala/com/github/tetrisanalyzer/piecegenerator/PieceGenerator.scala
package com.github.tetrisanalyzer.piecegenerator

import com.github.tetrisanalyzer.piece.Piece

trait PieceGenerator {
  def nextPieceNumber: Int
  def nextPiece = {
    val pieceNumber = nextPieceNumber
    require(pieceNumber >= 1 && pieceNumber <= 7, "Piece number must be in the range 1..7, found: " + pieceNumber)
    Piece(pieceNumber)
  }
}

File: src/main/java/com/github/tetrisanalyzer/piecegenerator/PieceGenerator.java
package nu.tengstrand.tetrisanalyzer.piecegenerator;

import nu.tengstrand.tetrisanalyzer.piece.Piece;

public abstract class PieceGenerator {
    public abstract int nextPieceNumber();

    public Piece nextPiece() {
        int pieceNumber = nextPieceNumber();

        if (pieceNumber < 1 || pieceNumber > 7) {
            throw new IllegalArgumentException("Piece number must be in the range 1..7, found: " + pieceNumber);
        }
        return Piece.get(pieceNumber);
    }
}
